text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,626 @@
+similar to the short-list of packages shown on the landing page.
+
+### Package Detail Page
+
+> [Example rendering](#rendering-package-detail-page)
+
+Clicking on a given package's tile in the landing page or search result page
+brings the user to a package detail page that represents the following
+information:
+
+- Package name, tags, and description
+- Overview of publishing information (author, version, license, etc...)
+  - Where possible, links to the source repository, issue tracker, etc... based
+    on information registered in the `.jsii` assembly file.
+- Install instructions
+- Available languages
+- Package's documentation
+  - Rendered content of the package's `README.md`
+  - API reference documentation, in all supported languages (only the version
+    for the currently selected language is displayed)
+
+The list of available languages is also a selector: upon clicking a language,
+the installation instructions and documentation automatically switch to the
+relevant form for the currently selected language.
+
+### Support
+
+The Construct Hub offers ways for customers to engage maintainers about
+various problems, including the following (actually routed to GitHub issue
+templates):
+
+- A malicious package was identified
+- Incorrectly rendered documentation
+- Package un-listing request
+
+When a contact request originates from a package detail page, the customer is
+guided to the package's issue tracker or public repository if their issue is
+with the package itself (and not about it's listing in Construct Hub). The
+routing is based on metadata configured on the package itself, through the bug
+tracker URL, repository URL, and homepage URL.
+
+## Technical Design
+
+### Overview
+
+![Architecture Diagram](../images/cdk-construct-hub/architecture-diagram.png)
+
+### What is the high level implementation plan?
+
+#### Overview
+
+The Construct Hub is a relatively simple static web application. While the
+dataset it exposes (all Construct Libraries) may be relatively fast moving,
+there is no necessity for newly published packages (or versions thereof) to be
+indexed and presented on the application particularly quickly. A consistency
+window between a couple of hours and a day is probably acceptable.
+
+A simple data pipeline will ingest package publication events created by a
+source-specific adapter (e.g: by polling on the CouchDB instance replica for the
+[npmjs.com registry](https://skimdb.npmjs.com/registry)), and produce static
+artifacts the website is based off of. This pipeline is in particular
+responsible for extracting the `.jsii` assembly file from the npm packages, and
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. **Package Discovery:** The back-end for Construct Hub starts with a custom
+   built event source, which implements the logic necessary to detect new
+   packages that are relevant to the Construct Hub, and sends messages to an SQS
```

Comment:
I think we should add some form of early filtering at the ingestion point. For example, we should skip any versions that we have already processed at this stage instead of queueing them, right?

Reply:
I am just looking for a way to identify filter construct libraries early, and wondered if a dependency on  CDK is a good signal

Reply:
It might be. But then also - what are we trying to protect against?